chain/client/src/sync/state/downloader.rs

```diff
@@ -69,13 +69,18 @@ impl StateSyncDownloader {
             let i = AtomicUsize::new(0); // for easier Rust async capture
             let attempt = || {
                 async {
+                    // We cannot assume that either source is infallible. We interleave attempts
+                    // to the available sources until one of them gives us the state successfully.
                     let source = if fallback_source.is_some()
                         && i.load(Ordering::Relaxed) >= num_attempts_before_fallback
                     {
+                        i.store(0, Ordering::Relaxed);
```

Comment:
Just confirming, we are saying we attempt to use the preferred_source `num_attempts_before_fallback` times, then try the fallback_source once before repeating the whole process again?

Reply:
Yeah, that's right. So effectively the num_attempts_before_fallback is configuring the ratio of preferred_source attempts to fallback_source attempts.